src/retry-strategy.ts

```diff
@@ -32,92 +46,106 @@ export class ExponentialBackoffRetryStrategy implements RetryStrategy {
+    cancel(): void {
+        this.pendingTimeouts.forEach( (timeout) => {
+            window.clearTimeout(timeout);
+            this.pendingTimeouts.delete(timeout)
+        });        
+    }
+
+    private requestMethodIsSafe(): boolean {
+        switch(this.requestMethod){
+            case 'GET':
+            case 'HEAD':
+            case 'OPTIONS':
+            case 'SUBSCRIBE':
+                return true;
+            default:
+                return false;
         }
     }
 
     private shouldRetry(error: Error): RetryStrategyResult {
-
         this.logger.verbose(`${this.constructor.name}:  Error received`, error);
         
         if(this.retryCount >= this.limit && this.limit > 0 ){
             this.logger.verbose(`${this.constructor.name}:  Retry count is over the maximum limit: ${this.limit}`);
             return new DoNotRetry(error);
         }
 
-        let retryable = this.isRetryable(error);
-        if(retryable.isRetryable){
-
-            if(retryable.backoffMillis){
-                this.retryCount += 1;
-                return new Retry(retryable.backoffMillis);
-            }
-            else{
-                this.currentBackoffMillis = this.calulateMillisToRetry();
-                this.retryCount += 1;
-            
-                this.logger.verbose(`${this.constructor.name}: Will attempt to retry in: ${this.currentBackoffMillis}`);
-                return new Retry(this.currentBackoffMillis)
-            }
-        }
+        if (error instanceof ErrorResponse && error.headers['Retry-After']){
+            return new Retry(parseInt(error.headers['Retry-After']) * 1000);
+        } 
 
-        else{
-            this.logger.verbose(`${this.constructor.name}: Error is not retryable`, error);
-            return new DoNotRetry(error);
+        if (error instanceof NetworkError || this.requestMethodIsSafe() || this.retryUnsafeRequests) {
+            return this.shouldSafeRetry(error);
         }
+    
+        this.logger.verbose(`${this.constructor.name}: Error is not retryable`, error);
+        return new DoNotRetry(error);
     }
 
-    attemptRetry(error: Error): Promise<any> {
-        return new Promise((resolve, reject) => {
-
-            let shouldRetry = this.shouldRetry(error);
+    private shouldSafeRetry(error: Error){
+        if(error instanceof NetworkError){
+            return new Retry(this.calulateMillisToRetry());
+        }
 
-            if(shouldRetry instanceof DoNotRetry){
-                reject(error);
+        if(error instanceof ErrorResponse) {
+            if(error.statusCode >= 500 && error.statusCode < 600){
+                return new Retry(this.calulateMillisToRetry());
             }
-            else if(shouldRetry instanceof Retry) {
-                window.setTimeout(resolve, shouldRetry.waitTimeMillis);
+            if(error.statusCode === 401){
+                return new Retry(0) //Token expired - can retry immediately
```

Comment:
I'm thinking to implement this naïvely for now, and then get back to it once I've got that merged. The PR is growing unwieldy as it is.